### DISCUSSION TITLE
ETCD Defragmentation - Cronjob vs Event-Driven Job

### DISCUSSION THREAD

**Discussion by ugur99** [etcd-io/etcd#14975] [Q&A] [answered]
Hi,

I'm looking for the best practices for defragmenting an etcd cluster. I've read the official documentation, but haven't seen any suggested methods to automate this process. I've seen some `cronjobs` or `systemd.timer` with `systemd.service` to defrag each instance one by one in a for loop or asynchronously respectively. 

I thought that starting to defragmentation from other etcd instances and lastly defragmenting the leader one would be safer if we can assume the defragmentation may not be complete in election_timeout time. Also, it would be useful to change the leader with `etcdctl move-leader` to the defragmentation completed etcd node. Such a cronjob running on the k8s is one option. 

What about instead of running such a defragmentation job as a cron, does it sound good to run it on an event-driven structure? When the [defragmentation alert](https://github.com/etcd-io/etcd/blob/6d27a42b7d6191da43d27930282de5b9e54ead7c/contrib/mixin/mixin.libsonnet#L242-L253) became active the job can be triggered. 

I believe running a defragmentation job regularly could make much more sense rather than running on an event-driven format, because depending on the db size that you configured; defragmentation can take longer to complete; but yes, at the end of the day, it would not cause any problem if you start from the other etcd instances except the leader one to defrag. 

I would welcome to hear any ideas/practices about it.

**Comment by serathius** [comment, marked as answer]
KISS - keep it stupid simple.

Cron job that avoid running when not needed based on some fragmentation threshold should work good enough.

**Reply by guettli** [reply]
There is now https://github.com/ahrtr/etcd-defrag which handles etcd defragmentation

**Comment by guettli** [comment]
I suggest to add an example script to the docs, and created a feature request: https://github.com/etcd-io/etcd/issues/15477
